src/NuGet.Core/NuGet.Packaging/FallbackPackagePathResolver.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using NuGet.Common;
+using NuGet.Packaging.Core;
+using NuGet.Versioning;
+
+namespace NuGet.Packaging
+{
+    public class FallbackPackagePathResolver
+    {
+        private readonly List<VersionFolderPathResolver> _pathResolvers;
+
+        /// <summary>
+        /// Creates a package folder path resolver that scans multiple folders to find a package.
+        /// </summary>
+        /// <param name="pathContext">NuGet paths loaded from NuGet.Config settings.</param>
+        public FallbackPackagePathResolver(INuGetPathContext pathContext)
+            : this(pathContext?.UserPackageFolder, pathContext?.FallbackPackageFolders)
+        {
+
+        }
+
+        public FallbackPackagePathResolver(string UserPackageFolder, IEnumerable<string> fallbackPackageFolders)
+        {
+            if (fallbackPackageFolders == null)
+            {
+                throw new ArgumentNullException(nameof(fallbackPackageFolders));
+            }
+
+            if (UserPackageFolder == null)
+            {
+                throw new ArgumentNullException(nameof(UserPackageFolder));
+            }
+
+            var packageFolders = new List<string>();
+
+            // The user's packages folder may not exist, this is expected if the fallback
+            // folders contain all packages.
+            if (Directory.Exists(UserPackageFolder))
+            {
+                packageFolders.Add(UserPackageFolder);
+            }
+
+            // All fallback folders must exist
+            foreach (var path in fallbackPackageFolders)
+            {
+                if (!Directory.Exists(path))
+                {
+                    var message = string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.FallbackFolderNotFound,
+                        path);
+
+                    throw new PackagingException(message);
+                }
+
+                packageFolders.Add(path);
+            }
+
+            // Create path resolvers for each source.
+            _pathResolvers = PathUtility.GetUniquePathsBasedOnOS(packageFolders)
```

Comment:
This is risky since some OS -> case sensitive FS mapping is not consistent. If we de-dupe in a case sensitive way, what is the negative impact on Windows?